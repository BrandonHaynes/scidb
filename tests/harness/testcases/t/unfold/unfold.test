# Any queries passed during setup phase do not require a 'SCIDB QUERY'
# line in the test.expected file.
--setup

--test
--start-query-logging
# The test.expected file requires that each line here be duplicated 
# in this manner, followed by its expected output.
# SCIDB QUERY : <test action line>
# {i} v,w
# {0} 0,1
# {1} 1,2
# {2} 2,3
# {3} 3,4
# {4} 4,5
# {5} 5,6
# {6} 6,7
# {7} 7,8
# {8} 8,9
# {9} 9,10

# Test integer input
unfold(apply(build(<v:int64>[i=0:9,3,0],i),w,i+1))

# Test double input
unfold(apply(build(<v:double>[i=0:9,3,0],i),w,i+0.5))

# Test input with empty values
unfold(apply(build(<v:double NULL>[i=0:5,1,1],NULL),w,i+0.5))

# Test input with more than two attributes
unfold(apply(apply(apply(build(<v:double>[i=0:9,3,0],i),w,i+0.5),x,w+0.25),y,x+0.125))

# Test sparse input
unfold(apply(filter(build(<w:double>[i=2:2,4,0],2),1),h,2*w))

# Test sparse input wiht more than two attributes
unfold(apply(apply(apply(filter(build(<v:double>[i=2:2,4,0],2),i),w,i+0.5),x,w+0.25),y,x+0.125))

# Test sparse input with larger array
store(build(<index:int64>[i=0:9,10,0],i*50),sparse_9a)
store(build(<val:double>[i=0:9,10,0], 2.718*i),sparse_9b)
store(join(sparse_9a,sparse_9b), sparse_9ab)
store(redimension(sparse_9ab, <val:double>[index=0:*, 50,0]),sparse_9result)
store(apply(sparse_9result,w,2*val), sr2attr9)
unfold(sr2attr9)

# Test sparse input with another array
store(build(<index:int64>[i=0:4,10,0],i*50),sparse_4a)
store(build(<val:double>[i=0:4,10,0], 2.718*i),sparse_4b)
store(join(sparse_4a,sparse_4b), sparse_4ab)
store(redimension(sparse_4ab, <val:double>[index=0:*, 50,0]),sparse_4result)
store(apply(sparse_4result,w,2*val), sr2attr4)
unfold(sr2attr4)

# Test case for sparse n-dimensional input array
unfold(apply(unfold(sr2attr4),w,3.5*val))

# Test for when input array has only one attribute
unfold(build(<v:int64>[i=0:9,3,0],i))

# Test case for datetime type
create array datetime1 <dt:datetime,dt2:datetime>[i=0:*,100,0]
--stop-query-logging
load(datetime1,'${TEST_DATA_DIR}/dates.txt')
--start-query-logging
unfold(datetime1)

# Test case where, for sparse input, the data do not occur
# at the chunk boundaries.
store(build(<index:int64>[i=0:9,10,0],i*55),sparse_5a)
store(build(<val:double>[i=0:9,10,0], 3.1416*i),sparse_5b)
store(join(sparse_5a,sparse_5b), sparse_5ab)
store(redimension(sparse_5ab, <val:double>[index=0:*, 50,0]),sparse_5result)
store(apply(sparse_5result,w,2*val), sr2attr5)
unfold(sr2attr5)

# Negative test for when attributes are not of the same type
--error --code=scidb::SCIDB_SE_OPERATOR::SCIDB_LE_ILLEGAL_OPERATION "unfold(apply(build(<v:int64>[i=0:9,3,0],i),w,i+0.5))"

# Test case for n-dimensional dense input array
unfold(apply(apply(build(<u:int64>[i=0:4,2,0,j=0:4,2,0],(i+1)*(i+1)),w,2*u+1),x,3*w+2))

# Test case for slice (added as consequence of #4316)
slice(unfold(apply(build(<a0:int64 NULL DEFAULT -3>[d0=0:9,4,2],d0),a1,d0+1)),d0,3)

# Test case for slice (added as consequence of #4316)
consume(slice(unfold(apply(build(<a0:int64 NULL DEFAULT -3>[d0=0:9,4,2],d0),a1,d0+1)),d0,3))

--stop-query-logging
--cleanup

# Any queries passed during cleanup phase do not require a 'SCIDB QUERY'
# line in the test.expected file.
remove(sparse_9a)
remove(sparse_9b)
remove(sparse_9ab)
remove(sparse_9result)
remove(sr2attr9)
remove(sparse_4a)
remove(sparse_4b)
remove(sparse_4ab)
remove(sparse_4result)
remove(sr2attr4)
remove(sparse_5a)
remove(sparse_5b)
remove(sparse_5ab)
remove(sparse_5result)
remove(sr2attr5)
remove(datetime1)

# The test.expected file must end with two newlines (newline followed
# by a blank line).  As an example, this file ends in that manner.

