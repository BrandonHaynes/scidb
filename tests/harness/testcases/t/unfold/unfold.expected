SCIDB QUERY : <unfold(apply(build(<v:int64>[i=0:9,3,0],i),w,i+1))>
{i,unfold_1} v
{0,0} 0
{0,1} 1
{1,0} 1
{1,1} 2
{2,0} 2
{2,1} 3
{3,0} 3
{3,1} 4
{4,0} 4
{4,1} 5
{5,0} 5
{5,1} 6
{6,0} 6
{6,1} 7
{7,0} 7
{7,1} 8
{8,0} 8
{8,1} 9
{9,0} 9
{9,1} 10

SCIDB QUERY : <unfold(apply(build(<v:double>[i=0:9,3,0],i),w,i+0.5))>
{i,unfold_1} v
{0,0} 0
{0,1} 0.5
{1,0} 1
{1,1} 1.5
{2,0} 2
{2,1} 2.5
{3,0} 3
{3,1} 3.5
{4,0} 4
{4,1} 4.5
{5,0} 5
{5,1} 5.5
{6,0} 6
{6,1} 6.5
{7,0} 7
{7,1} 7.5
{8,0} 8
{8,1} 8.5
{9,0} 9
{9,1} 9.5

SCIDB QUERY : <unfold(apply(build(<v:double NULL>[i=0:5,1,1],NULL),w,i+0.5))>
{i,unfold_1} v
{0,0} null
{0,1} 0.5
{1,0} null
{1,1} 1.5
{2,0} null
{2,1} 2.5
{3,0} null
{3,1} 3.5
{4,0} null
{4,1} 4.5
{5,0} null
{5,1} 5.5

SCIDB QUERY : <unfold(apply(apply(apply(build(<v:double>[i=0:9,3,0],i),w,i+0.5),x,w+0.25),y,x+0.125))>
{i,unfold_1} v
{0,0} 0
{0,1} 0.5
{0,2} 0.75
{0,3} 0.875
{1,0} 1
{1,1} 1.5
{1,2} 1.75
{1,3} 1.875
{2,0} 2
{2,1} 2.5
{2,2} 2.75
{2,3} 2.875
{3,0} 3
{3,1} 3.5
{3,2} 3.75
{3,3} 3.875
{4,0} 4
{4,1} 4.5
{4,2} 4.75
{4,3} 4.875
{5,0} 5
{5,1} 5.5
{5,2} 5.75
{5,3} 5.875
{6,0} 6
{6,1} 6.5
{6,2} 6.75
{6,3} 6.875
{7,0} 7
{7,1} 7.5
{7,2} 7.75
{7,3} 7.875
{8,0} 8
{8,1} 8.5
{8,2} 8.75
{8,3} 8.875
{9,0} 9
{9,1} 9.5
{9,2} 9.75
{9,3} 9.875

SCIDB QUERY : <unfold(apply(filter(build(<w:double>[i=2:2,4,0],2),1),h,2*w))>
{i,unfold_1} w
{2,0} 2
{2,1} 4

SCIDB QUERY : <unfold(apply(apply(apply(filter(build(<v:double>[i=2:2,4,0],2),i),w,i+0.5),x,w+0.25),y,x+0.125))>
{i,unfold_1} v
{2,0} 2
{2,1} 2.5
{2,2} 2.75
{2,3} 2.875

SCIDB QUERY : <store(build(<index:int64>[i=0:9,10,0],i*50),sparse_9a)>
{i} index
{0} 0
{1} 50
{2} 100
{3} 150
{4} 200
{5} 250
{6} 300
{7} 350
{8} 400
{9} 450

SCIDB QUERY : <store(build(<val:double>[i=0:9,10,0], 2.718*i),sparse_9b)>
{i} val
{0} 0
{1} 2.718
{2} 5.436
{3} 8.154
{4} 10.872
{5} 13.59
{6} 16.308
{7} 19.026
{8} 21.744
{9} 24.462

SCIDB QUERY : <store(join(sparse_9a,sparse_9b), sparse_9ab)>
{i} index,val
{0} 0,0
{1} 50,2.718
{2} 100,5.436
{3} 150,8.154
{4} 200,10.872
{5} 250,13.59
{6} 300,16.308
{7} 350,19.026
{8} 400,21.744
{9} 450,24.462

SCIDB QUERY : <store(redimension(sparse_9ab, <val:double>[index=0:*, 50,0]),sparse_9result)>
{index} val
{0} 0
{50} 2.718
{100} 5.436
{150} 8.154
{200} 10.872
{250} 13.59
{300} 16.308
{350} 19.026
{400} 21.744
{450} 24.462

SCIDB QUERY : <store(apply(sparse_9result,w,2*val), sr2attr9)>
{index} val,w
{0} 0,0
{50} 2.718,5.436
{100} 5.436,10.872
{150} 8.154,16.308
{200} 10.872,21.744
{250} 13.59,27.18
{300} 16.308,32.616
{350} 19.026,38.052
{400} 21.744,43.488
{450} 24.462,48.924

SCIDB QUERY : <unfold(sr2attr9)>
{index,unfold_1} val
{0,0} 0
{0,1} 0
{50,0} 2.718
{50,1} 5.436
{100,0} 5.436
{100,1} 10.872
{150,0} 8.154
{150,1} 16.308
{200,0} 10.872
{200,1} 21.744
{250,0} 13.59
{250,1} 27.18
{300,0} 16.308
{300,1} 32.616
{350,0} 19.026
{350,1} 38.052
{400,0} 21.744
{400,1} 43.488
{450,0} 24.462
{450,1} 48.924

SCIDB QUERY : <store(build(<index:int64>[i=0:4,10,0],i*50),sparse_4a)>
{i} index
{0} 0
{1} 50
{2} 100
{3} 150
{4} 200

SCIDB QUERY : <store(build(<val:double>[i=0:4,10,0], 2.718*i),sparse_4b)>
{i} val
{0} 0
{1} 2.718
{2} 5.436
{3} 8.154
{4} 10.872

SCIDB QUERY : <store(join(sparse_4a,sparse_4b), sparse_4ab)>
{i} index,val
{0} 0,0
{1} 50,2.718
{2} 100,5.436
{3} 150,8.154
{4} 200,10.872

SCIDB QUERY : <store(redimension(sparse_4ab, <val:double>[index=0:*, 50,0]),sparse_4result)>
{index} val
{0} 0
{50} 2.718
{100} 5.436
{150} 8.154
{200} 10.872

SCIDB QUERY : <store(apply(sparse_4result,w,2*val), sr2attr4)>
{index} val,w
{0} 0,0
{50} 2.718,5.436
{100} 5.436,10.872
{150} 8.154,16.308
{200} 10.872,21.744

SCIDB QUERY : <unfold(sr2attr4)>
{index,unfold_1} val
{0,0} 0
{0,1} 0
{50,0} 2.718
{50,1} 5.436
{100,0} 5.436
{100,1} 10.872
{150,0} 8.154
{150,1} 16.308
{200,0} 10.872
{200,1} 21.744

SCIDB QUERY : <unfold(apply(unfold(sr2attr4),w,3.5*val))>
{index,unfold_1,unfold_2} val
{0,0,0} 0
{0,0,1} 0
{0,1,0} 0
{0,1,1} 0
{50,0,0} 2.718
{50,0,1} 9.513
{50,1,0} 5.436
{50,1,1} 19.026
{100,0,0} 5.436
{100,0,1} 19.026
{100,1,0} 10.872
{100,1,1} 38.052
{150,0,0} 8.154
{150,0,1} 28.539
{150,1,0} 16.308
{150,1,1} 57.078
{200,0,0} 10.872
{200,0,1} 38.052
{200,1,0} 21.744
{200,1,1} 76.104

SCIDB QUERY : <unfold(build(<v:int64>[i=0:9,3,0],i))>
{i,unfold_1} v
{0,0} 0
{1,0} 1
{2,0} 2
{3,0} 3
{4,0} 4
{5,0} 5
{6,0} 6
{7,0} 7
{8,0} 8
{9,0} 9

SCIDB QUERY : <create array datetime1 <dt:datetime,dt2:datetime>[i=0:*,100,0]>
Query was executed successfully

{i} dt,dt2
{0} '2009-11-25 16:12:19','1970-01-01 00:00:00'
{1} '2009-11-25 16:12:19','1970-01-01 00:00:00'
{2} '2009-11-26 16:13:19','1970-01-01 00:00:00'
{3} '2009-11-26 16:13:19','1970-01-01 00:00:00'
{4} '2009-11-27 16:14:19','1970-01-01 00:00:00'
{5} '2009-11-27 16:14:19','1970-01-01 00:00:00'

SCIDB QUERY : <unfold(datetime1)>
{i,unfold_1} dt
{0,0} '2009-11-25 16:12:19'
{0,1} '1970-01-01 00:00:00'
{1,0} '2009-11-25 16:12:19'
{1,1} '1970-01-01 00:00:00'
{2,0} '2009-11-26 16:13:19'
{2,1} '1970-01-01 00:00:00'
{3,0} '2009-11-26 16:13:19'
{3,1} '1970-01-01 00:00:00'
{4,0} '2009-11-27 16:14:19'
{4,1} '1970-01-01 00:00:00'
{5,0} '2009-11-27 16:14:19'
{5,1} '1970-01-01 00:00:00'

SCIDB QUERY : <store(build(<index:int64>[i=0:9,10,0],i*55),sparse_5a)>
{i} index
{0} 0
{1} 55
{2} 110
{3} 165
{4} 220
{5} 275
{6} 330
{7} 385
{8} 440
{9} 495

SCIDB QUERY : <store(build(<val:double>[i=0:9,10,0], 3.1416*i),sparse_5b)>
{i} val
{0} 0
{1} 3.1416
{2} 6.2832
{3} 9.4248
{4} 12.5664
{5} 15.708
{6} 18.8496
{7} 21.9912
{8} 25.1328
{9} 28.2744

SCIDB QUERY : <store(join(sparse_5a,sparse_5b), sparse_5ab)>
{i} index,val
{0} 0,0
{1} 55,3.1416
{2} 110,6.2832
{3} 165,9.4248
{4} 220,12.5664
{5} 275,15.708
{6} 330,18.8496
{7} 385,21.9912
{8} 440,25.1328
{9} 495,28.2744

SCIDB QUERY : <store(redimension(sparse_5ab, <val:double>[index=0:*, 50,0]),sparse_5result)>
{index} val
{0} 0
{55} 3.1416
{110} 6.2832
{165} 9.4248
{220} 12.5664
{275} 15.708
{330} 18.8496
{385} 21.9912
{440} 25.1328
{495} 28.2744

SCIDB QUERY : <store(apply(sparse_5result,w,2*val), sr2attr5)>
{index} val,w
{0} 0,0
{55} 3.1416,6.2832
{110} 6.2832,12.5664
{165} 9.4248,18.8496
{220} 12.5664,25.1328
{275} 15.708,31.416
{330} 18.8496,37.6992
{385} 21.9912,43.9824
{440} 25.1328,50.2656
{495} 28.2744,56.5488

SCIDB QUERY : <unfold(sr2attr5)>
{index,unfold_1} val
{0,0} 0
{0,1} 0
{55,0} 3.1416
{55,1} 6.2832
{110,0} 6.2832
{110,1} 12.5664
{165,0} 9.4248
{165,1} 18.8496
{220,0} 12.5664
{220,1} 25.1328
{275,0} 15.708
{275,1} 31.416
{330,0} 18.8496
{330,1} 37.6992
{385,0} 21.9912
{385,1} 43.9824
{440,0} 25.1328
{440,1} 50.2656
{495,0} 28.2744
{495,1} 56.5488

SCIDB QUERY : <unfold(apply(build(<v:int64>[i=0:9,3,0],i),w,i+0.5))>
[An error expected at this place for the query "unfold(apply(build(<v:int64>[i=0:9,3,0],i),w,i+0.5))". And it failed with error code = scidb::SCIDB_SE_OPERATOR::SCIDB_LE_ILLEGAL_OPERATION. Expected error code = scidb::SCIDB_SE_OPERATOR::SCIDB_LE_ILLEGAL_OPERATION.]

SCIDB QUERY : <unfold(apply(apply(build(<u:int64>[i=0:4,2,0,j=0:4,2,0],(i+1)*(i+1)),w,2*u+1),x,3*w+2))>
{i,j,unfold_2} u
{0,0,0} 1
{0,0,1} 3
{0,0,2} 11
{0,1,0} 1
{0,1,1} 3
{0,1,2} 11
{1,0,0} 4
{1,0,1} 9
{1,0,2} 29
{1,1,0} 4
{1,1,1} 9
{1,1,2} 29
{0,2,0} 1
{0,2,1} 3
{0,2,2} 11
{0,3,0} 1
{0,3,1} 3
{0,3,2} 11
{1,2,0} 4
{1,2,1} 9
{1,2,2} 29
{1,3,0} 4
{1,3,1} 9
{1,3,2} 29
{0,4,0} 1
{0,4,1} 3
{0,4,2} 11
{1,4,0} 4
{1,4,1} 9
{1,4,2} 29
{2,0,0} 9
{2,0,1} 19
{2,0,2} 59
{2,1,0} 9
{2,1,1} 19
{2,1,2} 59
{3,0,0} 16
{3,0,1} 33
{3,0,2} 101
{3,1,0} 16
{3,1,1} 33
{3,1,2} 101
{2,2,0} 9
{2,2,1} 19
{2,2,2} 59
{2,3,0} 9
{2,3,1} 19
{2,3,2} 59
{3,2,0} 16
{3,2,1} 33
{3,2,2} 101
{3,3,0} 16
{3,3,1} 33
{3,3,2} 101
{2,4,0} 9
{2,4,1} 19
{2,4,2} 59
{3,4,0} 16
{3,4,1} 33
{3,4,2} 101
{4,0,0} 25
{4,0,1} 51
{4,0,2} 155
{4,1,0} 25
{4,1,1} 51
{4,1,2} 155
{4,2,0} 25
{4,2,1} 51
{4,2,2} 155
{4,3,0} 25
{4,3,1} 51
{4,3,2} 155
{4,4,0} 25
{4,4,1} 51
{4,4,2} 155

SCIDB QUERY : <slice(unfold(apply(build(<a0:int64 NULL DEFAULT -3>[d0=0:9,4,2],d0),a1,d0+1)),d0,3)>
{unfold_1} a0
{0} 3
{1} 4

SCIDB QUERY : <consume(slice(unfold(apply(build(<a0:int64 NULL DEFAULT -3>[d0=0:9,4,2],d0),a1,d0+1)),d0,3))>
Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

Query was executed successfully

